### PR TITLE
User login with authentication 

### PR DESCRIPTION
**Pull request by chriskevs**

### CODE REVIEW

File: source/backend/routes/auth.js [hunk 1]
```diff
@@ -1,10 +1,61 @@
...
+ * @param userId
+ * @returns {Promise<awaited Knex.QueryBuilder<TRecord, TResult>>}
+ */
+ async function getUserByUsername(username) {
+  const result = await db('users')
+    .select('username', 'userId', 'password')
+    .where({ username });
+  return result;
+}
+
```
Comment (demyinn00) [inline source/backend/routes/auth.js:19]: I think this should go into userModels.js?

File: source/backend/routes/auth.js [hunk 2]
```diff
@@ -1,10 +1,61 @@
...
+  if (rows.length === 0 || !match) {
+    return res.status(404).json({ message: 'User information not found' });
+  }
+  
+  // get userId and password
+  const user = { userId: rows[0]['userId'], password: rows[0]['password'] };
+
+  // return JWT token
+  const accessToken = jwt.sign(user, process.env.ACCESS_TOKEN_SECRET);
+  return res.status(200).json({ accessToken: accessToken });
```
Comment (demyinn00) [inline source/backend/routes/auth.js:58]: I think this looks pretty good, but I'm having a bit of trouble testing this. How were you able to test it? Does this require a user sign up first? In users, there is already a user object with a password with an encryption, but I don't know the password. How can I create a dummy user with a password will that require sign up ?

File: source/backend/routes/auth.js [hunk 3]
```diff
@@ -1,10 +1,61 @@
...
+  // determine if user exists in database
+  if (rows.length === 0 || !match) {
+    return res.status(404).json({ message: 'User information not found' });
+  }
+  
+  // get userId and password
+  const user = { userId: rows[0]['userId'], password: rows[0]['password'] };
+
+  // return JWT token
+  const accessToken = jwt.sign(user, process.env.ACCESS_TOKEN_SECRET);
```
Comment (mingkaili) [inline source/backend/routes/auth.js:57]: This function is already written for you in auth/tokenUtil.js it's called generateToken. 

All you have to do is: const accessToken = await tokenUtil.generateToken(user);

File: source/backend/routes/auth.js [hunk 4]
```diff
@@ -1,10 +1,61 @@
...
+       Return 404 (not found)
+*/
+router.post('/login', async (req, res) => {
+  const { username, password } = req.body;
+
+  // ensure user info is entered 
+  if (!username || !password) return res.status(400).json({ 'message': 'Username and password are requried.' });
+  
+  // find user
+  const rows = await getUserByUsername(username);
```
Comment (mingkaili) [inline source/backend/routes/auth.js:42]: move this function to usersModel as David mentioned above.

File: source/backend/routes/auth.js [hunk 5]
```diff
@@ -1,10 +1,61 @@
...
+  // evaluate passwords
+  const match = await bcrypt.compare(password, foundPassword);
+
+  // determine if user exists in database
+  if (rows.length === 0 || !match) {
+    return res.status(404).json({ message: 'User information not found' });
+  }
+  
+  // get userId and password
+  const user = { userId: rows[0]['userId'], password: rows[0]['password'] };
```
Comment (mingkaili) [inline source/backend/routes/auth.js:54]: Maybe a good idea to exclude the password in the token payload.

**Review by mingkaili — CHANGES_REQUESTED** [review]

**Pull request merged**